torchvision/csrc/io/image/cpu/decode_gif.cpp

```diff
@@ -0,0 +1,141 @@
+  // since we need to read the encoded bytes from a tensor of encoded bytes, not
+  // from a file (for consistency with existing jpeg and png decoders). Using
+  // DGifOpen is the only way to read from a custom source.
+  // For that we need to provide a reader function `read_from_tensor` that
+  // reads from the tensor, and we have to keep track of the number of bytes
+  // read so far: this is why we need the reader_helper struct.
+
+  // TODO: We are potentially doing an unnecessary copy of the encoded bytes:
+  // - 1 copy in from file to tensor (in read_file())
+  // - 1 copy from tensor to GIFLIB buffers (in read_from_tensor())
+  // Since we're vendoring GIFLIB we can potentially modify the calls to
+  // InternalRead() and just set the `buf` pointer to the tensor data directly.
+  // That might even save allocation of those buffers.
+  // If we do that, we'd have to make sure the buffers are never written to by
+  // GIFLIB, otherwise we'd be overridding the tensor data.
+  reader_helper_t reader_helper;
+  reader_helper.encoded_data = encoded_data.data_ptr<uint8_t>();
+  reader_helper.encoded_data_size = encoded_data.numel();
+  reader_helper.num_bytes_read = 0;
+  GifFileType* gifFile =
+      DGifOpen((void*)&reader_helper, read_from_tensor, &error);
+
+  TORCH_CHECK(
+      (gifFile != nullptr) && (error == D_GIF_SUCCEEDED),
+      "DGifOpenFileName() failed - ",
+      error);
+
+  if (DGifSlurp(gifFile) == GIF_ERROR) {
+    auto gifFileError = gifFile->Error;
+    DGifCloseFile(gifFile, &error);
+    TORCH_CHECK(false, "DGifSlurp() failed - ", gifFileError);
+  }
+
+  // Note:
+  // The GIF format has this notion of "canvas" and "canvas size", where each
+  // image could be displayed on the canvas at different offsets, forming a
+  // mosaic/picture wall like so:
+  //
+  // <---    canvas W    --->
+  // ------------------------     ^
+  // |         |            |     |
+  // |   img1  |    img3    |     |
+  // |         |------------|  canvas H
+  // |----------            |     |
+  // |   img2  |    img4    |     |
+  // |         |            |     |
+  // ------------------------     v
+  // The GifLib docs indicate that this is mostly vestigial, and modern viewers
+  // ignore the canvas size as well as image offsets. Hence, we're ignoring that
+  // too:
+  // - We're ignoring the canvas width and height and assume that the shape of
+  // the canvas and of all images is the shape of the first image.
+  // - We're enforcing that all images have the same shape.
+  // - Left and Top offsets of each image are ignored as well and assumed to be
+  // 0.
+
+  auto out_h = gifFile->SavedImages[0].ImageDesc.Height;
+  auto out_w = gifFile->SavedImages[0].ImageDesc.Width;
+  auto num_images = gifFile->ImageCount;
+
+  auto options = torch::TensorOptions()
+                     .dtype(torch::kU8)
+                     .memory_format(torch::MemoryFormat::ChannelsLast);
```

Comment:
I don't understand why this is ChannelsLast while below and in the loop the channels are before height and width?  I am likely missing something. Could you explain in a code comment so other readers can also understand better?

Reply:
As discussed offline the shape of the tensor is NCHW, but its format/layout is channels-last i.e. NHWC. Below we are using an accessor on the tensor data, whose indexing follows that of the shape i.e. `[n][c][h][w]` but the underlying operation still respects the channels-last format. If it didn't, I think our correctness tests would fail hard (and maybe even we'd have a segfault since `W >> C`?)

@vfdev-5 can you confirm that the accessor `out_a` properly respects the channels-last attribute?

Reply:
Yes, `accessor` is respecting tensor strides: https://github.com/pytorch/pytorch/blob/7864d287a1e56685aa754285cc2d3c31ff055f62/aten/src/ATen/core/TensorAccessor.h#L97-L109